### PR TITLE
Spheres performance test

### PR DESCRIPTION
**Pull request by CedricGuillemet**

### CODE REVIEW

File: Apps/UnitTests/Shared/Tests.h [hunk 1]
```diff
@@ -98,6 +98,91 @@ TEST(NativeAPI, LifeCycle)
...
+    runtime.Dispatch([&ready, &device, &nativeCanvas](Napi::Env env) {
+        device.AddToJavaScript(env);
+
+        Babylon::Polyfills::XMLHttpRequest::Initialize(env);
+        Babylon::Polyfills::Console::Initialize(env, [](const char* message, auto) {
+            printf("%s", message);
+            fflush(stdout);
+            });
+        Babylon::Polyfills::Window::Initialize(env);
+        nativeCanvas.emplace(Babylon::Polyfills::Canvas::Initialize(env));
```
Comment (bghgary) [inline Apps/UnitTests/Shared/Tests.h:154]: Why do you need canvas?

File: Apps/UnitTests/Shared/Tests.h [hunk 2]
```diff
@@ -98,6 +98,91 @@ TEST(NativeAPI, LifeCycle)
...
 */
+
+TEST(Performance, Spheres)
+{
+    // create a bunch of sphere, does the rendering for a number of frames, log time it took
+    static const std::string script{ R"(
+        console.log("Setting up Performance test.");
+        var engine = new BABYLON.NativeEngine();
+        var scene = new BABYLON.Scene(engine);
+        function CreateSpheresAsync(scene) {
```
Comment (bghgary) [inline Apps/UnitTests/Shared/Tests.h:109]: Creating spheres does not need to be async. It is only async in the experience to make the code consistent.

File: Apps/UnitTests/Shared/Tests.h [hunk 3]
```diff
@@ -98,6 +98,91 @@ TEST(NativeAPI, LifeCycle)
...
+            fflush(stdout);
+            });
+        Babylon::Polyfills::Window::Initialize(env);
+        nativeCanvas.emplace(Babylon::Polyfills::Canvas::Initialize(env));
+        Babylon::Plugins::NativeEngine::Initialize(env);
+        env.Global().Set("SetReady", Napi::Function::New(env, [&ready](const Napi::CallbackInfo& info)
+        {
+            Napi::Env env = info.Env();
+            ready.set_value(1);
+        }, "SetReady"));
```
Comment (bghgary) [inline Apps/UnitTests/Shared/Tests.h:160]: You don't need any of this stuff because creating spheres does not need to be async.

**Review by bghgary — CHANGES_REQUESTED** [review]

File: Apps/UnitTests/Shared/Tests.h (hunk 3 above)
Comment (CedricGuillemet) [reply Apps/UnitTests/Shared/Tests.h:160]: no but loading script is async. so, it's possible to finish rendering all frames while the spheres have not been created.

File: Apps/UnitTests/Shared/Tests.h (hunk 2 above)
Comment (CedricGuillemet) [reply Apps/UnitTests/Shared/Tests.h:109]: Fixed

File: Apps/UnitTests/Shared/Tests.h (hunk 3 above)
Comment (bghgary) [reply Apps/UnitTests/Shared/Tests.h:151]: Fair point.

File: Apps/UnitTests/Shared/Tests.h [hunk 4]
```diff
@@ -98,6 +98,82 @@ TEST(NativeAPI, LifeCycle)
...
+    runtime.Dispatch([&ready, &device](Napi::Env env) {
+        device.AddToJavaScript(env);
+
+        Babylon::Polyfills::Console::Initialize(env, [](const char* message, auto) {
+            printf("%s", message);
+            fflush(stdout);
+            });
+        Babylon::Polyfills::Window::Initialize(env);
+        Babylon::Plugins::NativeEngine::Initialize(env);
+        env.Global().Set("SetReady", Napi::Function::New(env, [&ready](const Napi::CallbackInfo& info)
```
Comment (bghgary) [inline Apps/UnitTests/Shared/Tests.h:147]: nit: use JS conventions

 ```suggestion
        env.Global().Set("setReady", Napi::Function::New(env, [&ready](const Napi::CallbackInfo& info)
```

File: Apps/UnitTests/Shared/Tests.h [hunk 5]
```diff
@@ -98,6 +98,82 @@ TEST(NativeAPI, LifeCycle)
...
+    {
+        device.StartRenderingCurrentFrame();
+        update->Start();
+        update->Finish();
+        device.FinishRenderingCurrentFrame();
+    }
+    // Stop measuring time
+    const auto stop = std::chrono::high_resolution_clock::now();
+    const auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(stop - start);
+    const float durationSeconds = float(double(duration.count()) / 1000.);
```
Comment (bghgary) [inline Apps/UnitTests/Shared/Tests.h:173]: nit: Why convert to double than back to float? Can you not convert to float directly?

File: Apps/UnitTests/Shared/Tests.h [hunk 6]
```diff
@@ -98,6 +98,82 @@ TEST(NativeAPI, LifeCycle)
...
+    printf("Duration is %f seconds.\n", durationSeconds);
```
Comment (bghgary) [inline Apps/UnitTests/Shared/Tests.h:174]: nit: Consider putting a flush after this also.

**Review by bghgary — APPROVED** [review]

File: Apps/UnitTests/Shared/Tests.h (hunk 5 above)
Comment (CedricGuillemet) [reply Apps/UnitTests/Shared/Tests.h:173]: I want to keep precision until the /1000.

Comment (bghgary) [reply Apps/UnitTests/Shared/Tests.h:173]: Divide by 1000 shouldn't change the precision much, but I'm okay if you want to keep it.

Comment (CedricGuillemet) [reply Apps/UnitTests/Shared/Tests.h:173]: Fixed

File: Apps/UnitTests/Shared/Tests.h (hunk 6 above)
Comment (CedricGuillemet) [reply Apps/UnitTests/Shared/Tests.h:174]: Fixed.

**Pull request merged**